riscv/triggers.cc

```diff
@@ -17,6 +17,94 @@ action_t trigger_t::legalize_action(reg_t val) const noexcept {
   return (val > ACTION_MAXVAL || (val == ACTION_DEBUG_MODE && get_dmode() == 0)) ? ACTION_DEBUG_EXCEPTION : (action_t)val;
 }
 
+reg_t trigger_t::tdata3_read(const processor_t * const proc) const noexcept {
+  auto xlen = proc->get_xlen();
+  reg_t tdata3 = 0;
+  if (xlen == 32) {
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_MHVALUE, mhvalue);
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_MHSELECT, mhselect);
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_SBYTEMASK, sbytemask);
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_SVALUE, svalue);
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_SSELECT, sselect);
+  } else {
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_MHVALUE, mhvalue);
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_MHSELECT, mhselect);
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_SBYTEMASK, sbytemask);
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_SVALUE, svalue);
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_SSELECT, sselect);
+  }
+  return tdata3;
+}
+
+void trigger_t::tdata3_write(processor_t * const proc, const reg_t val) noexcept {
+  auto xlen = proc->get_xlen();
+  if (xlen == 32) {
+    mhvalue = get_field(val, CSR_TEXTRA32_MHVALUE);
+    mhselect = get_field(val, CSR_TEXTRA32_MHSELECT);
+    sbytemask = get_field(val, CSR_TEXTRA32_SBYTEMASK);
+    svalue = get_field(val, CSR_TEXTRA32_SVALUE);
+    sselect = get_field(val, CSR_TEXTRA32_SSELECT);
+  } else {
+    mhvalue = get_field(val, CSR_TEXTRA64_MHVALUE);
+    mhselect = get_field(val, CSR_TEXTRA64_MHSELECT);
+    sbytemask = get_field(val, CSR_TEXTRA64_SBYTEMASK);
+    svalue = get_field(val, CSR_TEXTRA64_SVALUE);
+    sselect = get_field(val, CSR_TEXTRA64_SSELECT);
+  }
+  if (!proc->extension_enabled('H'))
+    mhselect = mhselect & MHSELECT_MCONTEXT; // the only legal value is 4, otherwise 0 (ignore)
+  if ((mhselect & MHSELECT_RESERVED) == MHSELECT_RESERVED) // 3 and 7 are reserved for mhselect
+    mhselect = 0;
```

Comment:
The WARLization logic here is pretty hard to understand and there are some embedded assumptions about the values of these enums (or else the bitwise logic wouldn't work). Since it's only a 3-bit field, can you express it as a table mapping the 8 possible inputs to 8 outputs?

Reply:
Add a legalization function ```legalize_mhselect()```.